model/float/riscv_float_eq.sail

```diff
@@ -0,0 +1,106 @@
+/*  1. Redistributions of source code must retain the above copyright                    */
+/*     notice, this list of conditions and the following disclaimer.                     */
+/*  2. Redistributions in binary form must reproduce the above copyright                 */
+/*     notice, this list of conditions and the following disclaimer in                   */
+/*     the documentation and/or other materials provided with the                        */
+/*     distribution.                                                                     */
+/*                                                                                       */
+/*  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''                   */
+/*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED                    */
+/*  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A                      */
+/*  PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR                  */
+/*  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,                         */
+/*  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT                     */
+/*  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF                     */
+/*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND                  */
+/*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,                   */
+/*  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT                   */
+/*  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF                   */
+/*  SUCH DAMAGE.                                                                         */
+/*=======================================================================================*/
+
+/* **************************************************************** */
+/* This file implements the floating-point for equal and consumes   */
+/* by the softfloat API contract, riscv_softfloat_interface.sail.   */
+/*                                                                  */
+/* The below floating-point types are supported.                    */
+/* 1. Double-precision, aka 64 bits floating-point.                 */
+/* 2. Single-precision, aka 32 bits floating-point.                 */
+/* 3. Half-precision, aka 16 bits floating-point.                   */
+/*                                                                  */
+/* **************************************************************** */
+
+val f_is_NaN      : forall 'm, 'm in {16, 32, 64}. bits('m) -> bool
+val f_is_SNaN     : forall 'm, 'm in {16, 32, 64}. bits('m) -> bool
+val accrue_fflags : (bits_fflags) -> unit
+
+val      riscv_float_eq : forall 'm, 'm in {16, 32, 64}. (bits('m), bits('m)) -> (bits_fflags, bool)
+function riscv_float_eq (op1, op2) -> (bits_fflags, bool) = {
+  let is_nan : bool = f_is_NaN(op1) | f_is_NaN(op2);
```

Comment:
You shouldn't need any of the explicit type annotations in this function. It can just be:

```
  let is_nan = f_is_NaN(op1) | f_is_NaN(op2);
  let is_snan = f_is_SNaN(op1) | f_is_SNaN(op2);
  let fflags = if is_nan & is_snan then float_flag_invalid else zeros();
  let is_equal = if not(is_nan)
                        then op1 == op2 | ((op1 | op2) << 1) == zeros()
                        else false;
```

Reply:
Not sure if I missed something, got some type/token error when removes the type annotations for the `is_equal`.

```sail
  27 ~ │   let is_equal = if   not(is_nan)
  28 ~ │                  then op1 == op2 | ((op1 | op2) << 1) == zeros()
  29 ~ │                  else false;
```

And then

```
Type error:
model/float/riscv_float_eq.sail:29.22-27:
29 |                 else false;
   |                      ^---^
   | Could not unify constraints ('ex2586# | 'ex2587#) and false
```